### ISSUE TITLE
Polymorphic TagSerializable

### ISSUE THREAD

**Issue by Chicken-Bones** [tModLoader/tModLoader#309]
The current pattern for polymorphic serialization using `TagSerializable` requires a `public new static Func<TagCompound, T> DESERIALIZER` field on every subtype. This is an unnecessary and counter-intuitive pattern for subclasses that don't change the serialization contract. The current `static DESERIALIZER` pattern was created as an attempt at "class level polymorphism" to allow classes to define their deserialisation method (normally call the appropriate constructor).

There is also a lack of documentation on correct use of the serialization system.
- `TagSerializable` should be used for your own classes
- `TagSerializer` should only be used for classes you don't own (eg, `Point` in another library) or structs

There are several possible solutions to the inheritance inadequacy
1. Let a `DESERIALIZER` field on a parent class function for retrieving instances of the children if not 'overriden'
    - This requires that the modder use `TagSerializer.GetType(tag.GetString("<type>"))` and call the appropriate constructor via reflection.
    - Should the `TagSerializableSerializer` enforce that the returned type is the same as `<type>` and how would that interact with refactoring (where `Get<T>` is called and `T` is not a parent of the type stored in `<type>` so `<type>` is ignored so the `DESERIALIZER` for `T` is called and allowed to re-interpret the data to match the requested type. (My intuition is that refactoring and inheritance don't mix under the current system unless the modder uses reflection and some kind of old -> new type mapping in their `DESERIALIZER`
2. Allow a constructor with a single `TagCompound` parameter to take the place of the `DESERIALIZER` field.
    - This should probably be implemented anyway. It has the downside that the constructor must be explicitly overriden in all subclasses, but this is much more intuitive.
    - The reason this option wasn't chosen in the first place is it doesn't allow for as much flexibility deserializer implementation, but it should probably be the default for most modders and documented as such.
3. Create an extension of `TagSerializable` like `TagSerializableAuto` (can't think of a good name at the time) which defines an additional `void Deserialize(TagCompound)` method, and calls the default constructor

**Comment by ChrisDirkis** [comment]
Not knowing the system, I like option 3 the most -- it requires no knowledge of the base class or requirements thereof (only that the auto class exists), and explicitly lays out the difference between the two classes (one class automatically does x, so the other clearly requires manual setup)

**Comment by Chicken-Bones** [comment]
Option 2 is best practice. If an object can be constructed from a TagCompound, it should have a constructor with a TagCompound parameter. The downside is every subclass has to override this constructor, but it is only a single line, and a common pattern. Languages require explicit constructor overrides so you can look at a class and see all the ways to construct an instance of it.
`public MySubclass(TagCompound tag) : base(tag) {}`

Option 3 is a matter of choice to do no-arg constructors and Init methods (which is a common pattern in serialization libraries, it just means you can't have readonly fields).

In summary:
1 is a grisly swiss army knife that someone might need somewhere sometime
2 encourages people to learn good practices
3 is something you should do when:
  A. people are novice programmers and you're not trying to teach code design (like most of our target audience
  B. you have one of those edge-case designs

**Comment by Jofairden** [comment]
Thing with option 2 is, even it being best practice, I'm just looking at it at low level: that's just as tedious as having to specify DESERIALIZER in every subclass so it only seems a marginal improvement at best.

**Comment by Rartrin** [comment]
We could make a system similar to the attribute one from mscorlib. This would follow closer to option 3 but with the ability of using attributes to allow for easier control of what should be serialized.
We would have ITagSerializableAuto or IAutoTaggable or something like that. Anything implementing the interface would indicate that we could use reflection and just manually save/load the fields in there by name, unless a field had a special attribute like [NotTaggable] or [NotTagSerializable], that would stop a field from being serialized. It would have to use the default constructor too for deserialization.
From there, we would have an another interface like ITagSerializeManual or something that extends IAutoTaggable for manual control of it. We would go ahead and just continue with the default constructor and then have the interface require instance methods for serial/deserializing.
By having them be interfaces, it means it will allow it to work for structs and such versus just a class.
We could also make it support the mcsorlib serializable if we wanted, that would mean we wouldn't have to hard code stuff for the primitive types.

**Comment by Chicken-Bones** [comment]
I approve, `[DontSerialize]` or `[NonSerialized]` `IAutoTagSerialized`

**Comment by Jofairden** [comment]
> I approve, [DontSerialize] or [NonSerialized] IAutoTagSerialized
What about [NonSerializedAttribute](https://msdn.microsoft.com/en-us/library/system.nonserializedattribute(v=vs.110).aspx) ?

**Comment by Jofairden** [comment]
@Chicken-Bones Is this something you still want to pickup ?
